cpp/src/join/full_join.cuh

```diff
@@ -0,0 +1,181 @@
+			thrust::make_counting_iterator(end_counter),
+			invalid_index_map.begin(),
+			unmatched_indices.begin(),
+			thrust::identity<index_type>()) -
+		unmatched_indices.begin();
+	unmatched_indices.resize(compacted_size);
+	return unmatched_indices;
+}
+
+/* --------------------------------------------------------------------------*/
+/**
+* @Synopsis  Expands a buffer's size if the requested size is greater that its
+* current size.
+*
+* @Param buffer Address of the buffer to expand
+* @Param buffer_capacity Memory allocated for buffer
+* @Param buffer_size Number of elements in the buffer
+* @Param expand_size Amount of extra elements to be pushed into the buffer
+* @tparam data_type The type of data associated with the buffer
+* @tparam size_type The data type used for size calculations
+*
+* @Returns  cudaSuccess upon successful completion of buffer expansion. Otherwise returns
+* the appropriate CUDA error code
+*/
+/* ----------------------------------------------------------------------------*/
+template <typename data_type, typename size_type>
+gdf_error expand_buffer(
+        data_type ** buffer,
+        size_type * const buffer_capacity,
+        const size_type buffer_size,
+        const size_type expand_size) {
+    size_type requested_size = buffer_size + expand_size;
+    //No need to proceed if the buffer can contain requested additional elements
+    if (*buffer_capacity >= requested_size) {
+        return GDF_SUCCESS;
+    }
+    data_type * new_buffer{nullptr};
+    data_type * old_buffer = *buffer;
+    CUDA_TRY( cudaMalloc(&new_buffer, requested_size*sizeof(data_type)) );
+    CUDA_TRY( cudaMemcpy(new_buffer, old_buffer, buffer_size*sizeof(data_type), cudaMemcpyDeviceToDevice) );
+    CUDA_TRY( cudaFree(old_buffer) );
+    *buffer = new_buffer;
+    *buffer_capacity = requested_size;
+
+    return GDF_SUCCESS;
+}
+
+/* --------------------------------------------------------------------------*/
+/**
+* @Synopsis  Adds indices that are missing in r_index_ptr at the ends and places
+* JoinNoneValue to the corresponding l_index_ptr.
+*
+* @Param l_index_ptr Address of the left indices
+* @Param r_index_ptr Address of the right indices
+* @Param index_capacity Amount of memory allocated for left and right indices
+* @Param index_size Number of left and right indices
+* @Param max_index_value The maximum value an index can have in r_index_ptr
+* @tparam index_type The type of data associated with index_ptr
+* @tparam size_type The data type used for size calculations
+*
+* @Returns  cudaSuccess upon successful completion of append call. Otherwise returns
+* the appropriate CUDA error code
+*/
+/* ----------------------------------------------------------------------------*/
+template <typename index_type, typename size_type>
+gdf_error append_full_join_indices(
+        index_type ** l_index_ptr,
+        index_type ** r_index_ptr,
+        size_type * const index_capacity,
+        size_type * const index_size,
```

Comment:
If this is a pointer to a scalar value, shouldn't you just pass a reference instead?